src/theory/quantifiers/sygus/sygus_utils.h

```diff
@@ -0,0 +1,116 @@
+/*! \file sygus_utils.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief generic sygus utilities
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__QUANTIFIERS__SYGUS__SYGUS_UTILS_H
+#define CVC4__THEORY__QUANTIFIERS__SYGUS__SYGUS_UTILS_H
+
+#include <vector>
+
+#include "expr/node.h"
+#include "expr/subs.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+class SygusUtils
+{
+ public:
+  /**
+   * Make (negated) sygus conjecture of the form
+   *   forall fs. conj
+   * with instantiation attributes in iattrs. Notice that the marker for
+   * sygus conjecture is automatically prepended to this list.
+   *
+   * @param fs The functions
+   * @param conj The (negated) conjecture body
+   * @param iattrs The attributes of the conjecture. Notice this does not
+   * require the "sygus attribute" marker, which is automatically generated
+   * by this method.
+   */
+  static Node mkSygusConjecture(const std::vector<Node>& fs,
+                                Node conj,
+                                const std::vector<Node>& iattrs);
+  /** Same as above, without auxiliary instantiation attributes */
+  static Node mkSygusConjecture(const std::vector<Node>& fs, Node conj);
+
+  /**
+   * Make conjecture, with a set of solved functions. In particular,
+   * solvedf is a substitution of the form { f1 -> t1, ... fn -> tn } where
+   * each f1 ... fn are in fs.
+   *
+   * In the implementation, solutions for solved functions are stored
+   * in the instantiation attribute list of the returned conjecture.
+   */
+  static Node mkSygusConjecture(const std::vector<Node>& fs,
+                                Node conj,
+                                const Subs& solvedf);
+  /**
+   * Decompose sygus conjecture.
+   *
+   * @param q The sygus conjecture to decompose
+   * @param fs The functions-to-synthesize
+   * @param unsf The functions that have not been marked as solved.
+   * @param solvedf The substitution corresponding to the solved functions.
+   *
+   * The vector unsf and the domain of solved are a parition of fs.
+   */
+  static void decomposeSygusConjecture(Node q,
+                                       std::vector<Node>& fs,
+                                       std::vector<Node>& unsf,
+                                       Subs& solvedf);
+  /**
+   * Decompose the negated conjecture body.
+   *
+   * This returns a quantifier-free formula corresponding to the (un-negated)
+   * conjecture body. It adds the quantified free variables of the conjecture
+   * to vs.
+   */
+  static Node decomposeConjectureBody(Node conj, std::vector<Node>& vs);
```

Comment:
It's a bit confusing to have this and the previous method named similarly but doing seemly very different things.

Reply:
Maybe `decomposeSygusConjecture` and `decomposeSygusBody` ?

Reply:
I think what was more prominent was that the first decompose the functions-to-synthesize and the second gets the vars and *negates* the body. Having the fact the body is negated in the name would be helpful, I think. Maybe something like `getNegatedConjectureBodyAndFVs`?

Reply:
I think that's a bit verbose :) decomposeSygusNegatedBody?

Reply:
That sounds good!

Reply:
Decided to change the negation in the method, since it was confusing, and call this `decomposeSygusBody` since it is now analogous to the other method (operating on negated conjectures/their bodies).